spec/lib/administrate/page/collection_spec.rb

```diff
@@ -0,0 +1,399 @@
+require "spec_helper"
+require "support/constant_helpers"
+require "administrate/page/collection"
+require "administrate/search"
+
+describe Administrate::Page::Collection do
+  # Constants defined in spec_helper.rb
+  let(:scopes_array) { DashboardWithAnArrayOfScopes::COLLECTION_SCOPES }
+  let(:scopes_hash) { DashboardWithAHashOfScopes::COLLECTION_SCOPES }
+  let(:dashboard_wo_scopes) { MockDashboard.new }
+  let(:dashboard_w_scopes_array) { DashboardWithAnArrayOfScopes.new }
+  let(:dashboard_w_scopes_hash) { DashboardWithAHashOfScopes.new }
+
+  # #scope_groups creates the concept of "group of scopes" to manage scopes
+  # always grouped reading Dashboard#collection_scopes (COLLECTION_SCOPES).
+  describe "#scope_groups" do
+    describe "with no scopes defined" do
+      it "returns an empty array" do
+        page = Administrate::Page::Collection.new(dashboard_wo_scopes)
+        expect(page.scope_groups).to eq([])
+      end
+    end
+
+    describe "with an Array of scopes" do
+      it "returns an array with the :scopes symbol inside ([:scopes])" do
+        page = Administrate::Page::Collection.new(dashboard_w_scopes_array)
+        expect(page.scope_groups).to eq([:scopes])
+      end
+    end
+
+    describe "with a Hash grouping the scopes" do
+      it "returns the hash's keys" do
+        page = Administrate::Page::Collection.new(dashboard_w_scopes_hash)
+        expect(page.scope_groups).to eq(scopes_hash.keys)
+      end
+    end
+  end
+
+  # #scope_names([group]) returns an array with *scope names*. A *scope name*
+  # can be a symbol or a string that matchs with an scope defined in the
+  # dashboard's model including the scope's argument when needed (e.g.
+  # ["valid", "awesome_since(2004)"]).
+  describe "#scope_names([group])" do
+    describe "with no scopes defined" do
+      it "returns an empty array (and ignores group passed as argument)" do
+        page = Administrate::Page::Collection.new(dashboard_wo_scopes)
+        expect(page.scope_names).to eq([])
+        expect(page.scope_names(:scopes)).to eq([])
+      end
+    end
+
+    describe "with an Array of scopes" do
+      let(:array_of_strings) { scopes_array.map(&:to_s) }
+
+      it "returns that array (and ignores group passed as argument)" do
+        page = Administrate::Page::Collection.new(dashboard_w_scopes_array)
+        expect(page.scope_names).to eq(array_of_strings)
+        expect(page.scope_names(:scopes)).to eq(array_of_strings)
+      end
+    end
+
+    describe "with a Hash grouping the scopes" do
+      let(:scopes_strigified_without_wildcarded) do
+        scopes_hash[:status].map(&:to_s).reject { |s| s[-2..-1] == ':*'}
```

Comment:
Space missing inside }.<br>Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.